protocol/contracts/zaps/GenericZap.sol

```diff
@@ -0,0 +1,697 @@
+
+    // pull tokens
+    _pullTokens(_fromToken, _fromAmount);
+
+    address intermediateToken;
+    uint256 intermediateAmount;
+    (address token0, address token1) = Swap.getPairTokens(_pair);
+
+    if (_fromToken != token0 && _fromToken != token1) {
+      // swap to intermediate
+      intermediateToken = _zapLiqRequest.otherToken == token0 ? token1 : token0;
+      intermediateAmount = Swap.fillQuote(
+        _fromToken,
+        _fromAmount,
+        intermediateToken,
+        _swapTarget,
+        _swapData
+      );
+      require(intermediateAmount >= _zapLiqRequest.firstSwapMinAmountOut, "Not enough tokens out");
+    } else {
+      intermediateToken = _fromToken;
+      intermediateAmount = _fromAmount;
+    }
+    
+    (uint256 amountA, uint256 amountB) = _swapTokens(_pair, intermediateToken, intermediateAmount, _zapLiqRequest.poolSwapMinAmountOut);
+
+    _approveToken(token1, address(uniswapV2Router), amountB);
+    _approveToken(token0, address(uniswapV2Router), amountA);
+
+    _addLiquidityUniV2(_pair, _for, amountA, amountB, _zapLiqRequest.uniAmountAMin, _zapLiqRequest.uniAmountBMin, _zapLiqRequest.shouldTransferResidual);
+  }
+
+  /**
+   * @dev Get minimum amounts fo token0 and token1 to tolerate when adding liquidity to uniswap v2 pair
+   * @param amountADesired Input desired amount of token0
+   * @param amountBDesired Input desired amount of token1
+   * @param pair Target uniswap v2 pair
+   * @return amountA Minimum amount of token0 to use when adding liquidity
+   * @return amountB Minimum amount of token1 to use when adding liquidity
+   */
+  function addLiquidityGetMinAmounts(
+    uint amountADesired,
+    uint amountBDesired,
+    IUniswapV2Pair pair
+  ) public view returns (uint amountA, uint amountB) {
+    (uint reserveA, uint reserveB,) = pair.getReserves();
+    if (reserveA == 0 && reserveB == 0) {
+      (amountA, amountB) = (amountADesired, amountBDesired);
+    } else {
+      uint amountBOptimal = Swap.quote(amountADesired, reserveA, reserveB);
+      if (amountBOptimal <= amountBDesired) {
+          //require(amountBOptimal >= amountBMin, 'TempleStableAMMRouter: INSUFFICIENT_STABLE');
+          (amountA, amountB) = (amountADesired, amountBOptimal);
+      } else {
+          uint amountAOptimal = Swap.quote(amountBDesired, reserveB, reserveA);
+          assert(amountAOptimal <= amountADesired);
+          //require(amountAOptimal >= amountAMin, 'TempleStableAMMRouter: INSUFFICIENT_TEMPLE');
+          (amountA, amountB) = (amountAOptimal, amountBDesired);
+      }
+    }
+  }
+
+  function _addLiquidityCurvePool(
+    address _pool,
+    address _recipient,
+    uint256 _minLiquidityOut,
+    bool _useAltFunction,
+    uint256[] memory _amounts
+  ) internal returns (uint256 liquidity) {
+    bool success;
+    bytes memory data;
+    if (_useAltFunction) {
+      //liquidity = ICurvePool(_pool).add_liquidity(_amounts, _minLiquidityOut, false, _recipient);
```

Comment:
Specifying the direct hex to save the `bytes4(keccak256("add_liquidity.....))` call? What's the extra bool param?

Reply:
1/ yes.
2/ the extra bool param is a special case for some base 2 plain pools eg. fxsCvxFxs pool
I've added this explanation to documentation. will push commit soon